execution/function/operator.go

```diff
@@ -33,6 +33,96 @@ type functionOperator struct {
 	pointBuf     []promql.Point
 }
 
+type absentOperator struct {
+	mint        int64
+	maxt        int64
+	step        int64
+	currentStep int64
+	stepsBatch  int
+	funcExpr    *parser.Call
+	series      []labels.Labels
+	once        sync.Once
+
+	next       model.VectorOperator
+	vectorPool *model.VectorPool
+}
+
+func (o *absentOperator) Explain() (me string, next []model.VectorOperator) {
+	return fmt.Sprintf("[*absentOperator] %v()", o.funcExpr.Func.Name), []model.VectorOperator{}
+}
+
+func (o *absentOperator) Series(_ context.Context) ([]labels.Labels, error) {
+	o.loadSeries()
+
+	return o.series, nil
+}
+
+func (o *absentOperator) loadSeries() {
+	o.once.Do(func() {
+		// https://github.com/prometheus/prometheus/blob/main/promql/functions.go#L1385
+		var lm []*labels.Matcher
+		switch n := o.funcExpr.Args[0].(type) {
+		case *parser.VectorSelector:
+			lm = n.LabelMatchers
+		case *parser.MatrixSelector:
+			lm = n.VectorSelector.(*parser.VectorSelector).LabelMatchers
+		default:
+			o.series = []labels.Labels{labels.EmptyLabels()}
+			return
+		}
+
+		has := make(map[string]bool)
+		lmap := make(map[string]string)
+		for _, l := range lm {
+			if l.Name == labels.MetricName {
+				continue
+			}
+			if l.Type == labels.MatchEqual && !has[l.Name] {
+				lmap[l.Name] = l.Value
+				has[l.Name] = true
+			} else {
+				delete(lmap, l.Name)
+			}
+		}
+		o.series = []labels.Labels{labels.FromMap(lmap)}
+	})
+}
+
+func (o *absentOperator) GetPool() *model.VectorPool {
+	return o.vectorPool
+}
+
+func (o *absentOperator) Next(ctx context.Context) ([]model.StepVector, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+
+	if o.currentStep > o.maxt {
+		return nil, nil
+	}
+
+	vectors, err := o.next.Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(vectors) > 0 {
+		return nil, nil
```

Comment:
This in theory should cause the entire execution to stop. So if a series is present and then goes absent, we will never emit a `1` value. Is this how Prometheus works too?

Reply:
Ah really good point, 
```
		{
			name: "absent",
			load: `load 30s
				existent{job="myjob"} 1 1 1`,
			query: `absent(existent{job="myjob"})`,
		},
```
in `TestQueriesAgainstOldEngine` is failing exactely as you describe 

```
                    Diff:
                    --- Expected
                    +++ Actual
                    @@ -1,6 +1,2 @@
                    -(*promql.Result)({job="myjob"} =>
                    -1 @[150000]
                    -1 @[180000]
                    -1 @[210000]
                    -1 @[240000])
                    +(*promql.Result)()
```

Reply:
I assume the same would apply for the individual batches, we probably need to check on a per-step basis whether the metric is present or not.

Reply:
Yeah i think we need to step through the vector at timestamps and if we have no samples we need to add a sample for absent; if we hvae samples we need to drop them. Or something like that

Reply:
Ok i fixed the test in the latest version!